x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/validation/SourceDestValidator.java

```diff
@@ -0,0 +1,513 @@
+     *
+     * @param clusterState The current ClusterState
+     * @param indexNameExpressionResolver A valid IndexNameExpressionResolver object
+     * @param remoteClusterService A valid RemoteClusterService object
+     * @param remoteClusterLicenseChecker A RemoteClusterLicenseChecker or null if CCS is disabled
+     * @param source an array of source indices
+     * @param dest destination index
+     * @param nodeName the name of this node
+     * @param license the license of the feature validated for
+     * @return ValidationException instance containing all validation errors or null if validation found no problems.
+     */
+    public static ValidationException validateForPreview(
+        ClusterState clusterState,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        RemoteClusterService remoteClusterService,
+        RemoteClusterLicenseChecker remoteClusterLicenseChecker,
+        String[] source,
+        String dest,
+        String nodeName,
+        String license
+    ) {
+        Context context = new Context(
+            clusterState,
+            indexNameExpressionResolver,
+            remoteClusterService,
+            remoteClusterLicenseChecker,
+            source,
+            dest,
+            nodeName,
+            license
+        );
+
+        // validation exception might gets thrown if validation stops early
+        try {
+            for (SourceDestValidation validation : PREVIEW_VALIDATIONS) {
+                validation.validate(context);
+            }
+        } catch (ValidationException e) {}
+
+        return context.getValidationException();
+    }
+
+    /**
+     * Validates source and destination indices for put, start and update.
+     * Assumes that name checks have been made at the REST layer.
+     *
+     * @param clusterState The current ClusterState
+     * @param indexNameExpressionResolver A valid IndexNameExpressionResolver object
+     * @param remoteClusterService A valid RemoteClusterService object
+     * @param remoteClusterLicenseChecker A RemoteClusterLicenseChecker or null if CCS is disabled
+     * @param source an array of source indexes
+     * @param dest destination index
+     * @param nodeName the name of this node
+     * @param license the license of the feature validated for
+     * @param shouldDefer whether to defer certain validations, used for put while source indices are not created yet
+     * @return ValidationException instance containing all validation errors or null if validation found no problems.
+     */
+
+    public static ValidationException validate(
+        ClusterState clusterState,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        RemoteClusterService remoteClusterService,
+        RemoteClusterLicenseChecker remoteClusterLicenseChecker,
+        String[] source,
+        String dest,
+        String nodeName,
+        String license,
+        boolean shouldDefer
```

Comment:
I think it would be nice to pass in what validations the user wants to run. We could have all the available validations as public static values and then they pass in a variadic list of ValidationClasses or something.